numpy/core/src/multiarray/ctors.c

```diff
@@ -1238,8 +1241,8 @@ PyArray_NewLikeArrayWithShape(PyArrayObject *prototype, NPY_ORDER order,
             break;
     }
 
-    /* If it's not KEEPORDER, this is simple */
-    if (order != NPY_KEEPORDER) {
+    /* If it's not KEEPORDER, or there is a shape change, this is simple */
+    if (order != NPY_KEEPORDER || new_shape) {
```

Comment:
But passing shape alone doesn't provide any information of strides, it will not be more than an assumption that keeping strides unchanged is important. Plus, we could only keep the strides for the dimensions that don't change.

Reply:
The stride information comes from the source. The purpose of `KEEPORDER` is to preserve `np.argsort(arr.strides)`.

I think that the behavior should be:
* Same `ndim` - respect the stride ordering through keeporder
* different `ndim` - reject keeporder since it is meaningless, throw an exception asking the use to use corder or Forder.

Reply:
But even if `ndim` remains the same, suppose source `shape=(5, 2)` and new `shape=(10000, 10)`, in such a case we can't respect strides. Another possibility is source `shape=(10000, 10)` and new `shape=(5, 2)`, we could respect the strides at the cost of huge memory waste, which is most likely going to be an unexpected behavior from the user.

Reply:
keeporder means keep the relative order of the strides (`np.argsort(arr.strides)`), not the strides themselves (`arr.strides`).

Reply:
Oh, I'm sorry, I got what you mean now. Yes, I agree with that, let me change that.

Reply:
I think the new commits should cover this case now.